gateway-js/src/loadCsdlFromStorage.ts

```diff
@@ -0,0 +1,100 @@
+import { fetch } from 'apollo-server-env';
+import { GraphQLError } from 'graphql';
+import { CsdlQuery } from './__generated__/graphqlTypes';
+
+export const CSDL_QUERY = /* GraphQL */`#graphql
+  query Csdl($apiKey: String!, $ref: String!) {
+    routerConfig(ref: $ref, apiKey: $apiKey) {
+      __typename
+      ... on RouterConfigResult {
+        id
+        csdl
+      }
+      ... on FetchError {
+        code
+        message
+      }
+    }
+  }
+`;
+
+type CsdlQueryResult = CsdlQuerySuccess | CsdlQueryFailure;
+
+interface CsdlQuerySuccess {
+  data: CsdlQuery;
+}
+
+interface CsdlQueryFailure {
+  data?: CsdlQuery;
+  errors: GraphQLError[];
+}
+
+const { name, version } = require('../package.json');
+
+export async function loadCsdlFromStorage({
+  graphId,
+  graphVariant,
+  apiKey,
+  endpoint,
+  fetcher,
+}: {
+  graphId: string;
+  graphVariant: string;
+  apiKey: string;
+  endpoint: string;
+  fetcher: typeof fetch;
+}) {
+  const result = await fetcher(endpoint, {
+    method: 'POST',
+    body: JSON.stringify({
+      query: CSDL_QUERY,
+      variables: {
+        ref: `${graphId}@${graphVariant}`,
+        apiKey,
+      },
+    }),
+    headers: {
+      'apollographql-client-name': name,
+      'apollographql-client-version': version,
+      'user-agent': `${name}/${version}`,
+      'content-type': 'application/json',
+    },
+  });
+
+  let response: CsdlQueryResult;
+
+  if (result.ok || result.status === 400) {
+    try {
+      response = await result.json();
+    } catch (e) {
+      // Bad response
+      throw new Error(result.status + ': ' + e.message ?? e);
```

Comment:
This error seems pretty minimalist. Does some caller add more context to it explaining that the problem is in fetching the schema from Apollo? I don't think so on a quick glance. Ditto for the other two errors in the next few lines... and `await fetcher`...  That's what I was getting at with https://github.com/apollographql/federation/pull/458#discussion_r593522095 ... though I also said it could be dealt with later, are you tracking it somewhere else instead? (Could use TODO comments at least) But yeah, maybe this whole function could be in a try/catch/rethrow that adds some context, or its caller could do that?

Reply:
Let me know what you think about this. Still feels weird to me to just append a message but this seems pretty reasonable and with the added consistency we can know that an error with this message is always coming from our schema fetcher.
9b2e84e